<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Real world example: GB1 landscape &mdash; Visualization of fitness landscapes 0.1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="ParD antitoxin empirical landscape" href="3_ParD.html" />
    <link rel="prev" title="Getting started: serine landscape" href="1_Ser.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> Visualization of fitness landscapes
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation Instructions</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../usage.html">Usage examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="1_Ser.html">Getting started: serine landscape</a></li>
<li class="toctree-l2"><a class="reference internal" href="1_Ser.html#How-many-dimensions-to-plot?">How many dimensions to plot?</a></li>
<li class="toctree-l2"><a class="reference internal" href="1_Ser.html#Characterizing-fitness-peaks">Characterizing fitness peaks</a></li>
<li class="toctree-l2"><a class="reference internal" href="1_Ser.html#Setting-the-Ns-parameter">Setting the <span class="math notranslate nohighlight">\(Ns\)</span> parameter</a></li>
<li class="toctree-l2"><a class="reference internal" href="1_Ser.html#Advanced-landscape-plotting-and-customization">Advanced landscape plotting and customization</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Real world example: GB1 landscape</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#1.-Define-sequence-space">1. Define sequence space</a></li>
<li class="toctree-l3"><a class="reference internal" href="#2.-Define-random-walk-on-the-discrete-space">2. Define random walk on the discrete space</a></li>
<li class="toctree-l3"><a class="reference internal" href="#3.-Calculalate-visualization-coordinates">3. Calculalate visualization coordinates</a></li>
<li class="toctree-l3"><a class="reference internal" href="#4.-Plot-landscape">4. Plot landscape</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Characterize-the-fitness-peaks">Characterize the fitness peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Characterize-the-intermediates-between-peaks">Characterize the intermediates between peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Exploring-more-diffusion-axis">Exploring more diffusion axis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#References">References</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="3_ParD.html">ParD antitoxin empirical landscape</a></li>
<li class="toctree-l2"><a class="reference internal" href="4_SeqDEFT.html">Sequence Density Estimation using Field Theory</a></li>
<li class="toctree-l2"><a class="reference internal" href="5_Codon.html">Codon models</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Visualization of fitness landscapes</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../usage.html">Usage examples</a> &raquo;</li>
      <li>Real world example: GB1 landscape</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/usage/2_GB1.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="Real-world-example:-GB1-landscape">
<h1>Real world example: GB1 landscape<a class="headerlink" href="#Real-world-example:-GB1-landscape" title="Permalink to this headline"></a></h1>
<p>In the previous tutorial, we have seen how to create a series of plots for the very simple case of the serine landscape that was artificially built for this purpose and had only 64 genotypes. Therefore, we knew beforehand the qualitative properties of the landscape, and saw how they were reflected on the visualization. However, we usually would want to do the opposite process, in which we visualize an unknown landscape and want to understand what are its main properties.</p>
<p>In the followin tutorial we will show how to do that with a real world example using data from {cite}<code class="docutils literal notranslate"><span class="pre">Wu2016</span></code> for which we generated a visualization in previous work {cite}<code class="docutils literal notranslate"><span class="pre">Zhou2020c</span></code>. Now we can generate it and explore the different peaks in a few lines of python code.</p>
<p>As we have seen in previous notebooks, the visualizations can be easily created in few simple steps</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import required libraries</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>

<span class="kn">import</span> <span class="nn">gpmap.src.plot</span> <span class="k">as</span> <span class="nn">plot</span>

<span class="kn">from</span> <span class="nn">gpmap.src.inference</span> <span class="kn">import</span> <span class="n">VCregression</span>
<span class="kn">from</span> <span class="nn">gpmap.src.space</span> <span class="kn">import</span> <span class="n">SequenceSpace</span>
<span class="kn">from</span> <span class="nn">gpmap.src.randwalk</span> <span class="kn">import</span> <span class="n">WMWSWalk</span>
</pre></div>
</div>
</div>
<section id="1.-Define-sequence-space">
<h2>1. Define sequence space<a class="headerlink" href="#1.-Define-sequence-space" title="Permalink to this headline"></a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;../data/gb1.data.csv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">space</span> <span class="o">=</span> <span class="n">SequenceSpace</span><span class="p">(</span><span class="n">X</span><span class="o">=</span><span class="n">data</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;log_binding&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">space</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Sequence Space:
        Type: protein
        Sequence length: 4
        Number of alleles per site: [20, 20, 20, 20]
        Genotypes: [AAAA,AAAC,AAAD,...,YYYV,YYYW,YYYY]
        Function y: [0.30,-2.71,-2.91,...,-3.00,-4.72,-4.88]
</pre></div></div>
</div>
</section>
<section id="2.-Define-random-walk-on-the-discrete-space">
<h2>2. Define random walk on the discrete space<a class="headerlink" href="#2.-Define-random-walk-on-the-discrete-space" title="Permalink to this headline"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rw</span> <span class="o">=</span> <span class="n">WMWSWalk</span><span class="p">(</span><span class="n">space</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="3.-Calculalate-visualization-coordinates">
<h2>3. Calculalate visualization coordinates<a class="headerlink" href="#3.-Calculalate-visualization-coordinates" title="Permalink to this headline"></a></h2>
<p>In this particular case, the values are scaled so that the mean value for the WT sequences is 0. Then it seems a reasonable guess to set the mean function at stationarity at that value to choose <code class="docutils literal notranslate"><span class="pre">Ns</span></code></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rw</span><span class="o">.</span><span class="n">calc_visualization</span><span class="p">(</span><span class="n">mean_function</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">n_components</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">nodes_df</span><span class="p">,</span> <span class="n">edges_df</span> <span class="o">=</span> <span class="n">rw</span><span class="o">.</span><span class="n">nodes_df</span><span class="p">,</span> <span class="n">rw</span><span class="o">.</span><span class="n">space</span><span class="o">.</span><span class="n">get_edges_df</span><span class="p">()</span>
</pre></div>
</div>
</div>
</section>
<section id="4.-Plot-landscape">
<h2>4. Plot landscape<a class="headerlink" href="#4.-Plot-landscape" title="Permalink to this headline"></a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plot</span><span class="o">.</span><span class="n">init_fig</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">plot</span><span class="o">.</span><span class="n">plot_relaxation_times</span><span class="p">(</span><span class="n">rw</span><span class="o">.</span><span class="n">decay_rates_df</span><span class="p">,</span> <span class="n">axes</span><span class="p">,</span> <span class="n">neutral_time</span><span class="o">=</span><span class="mi">1</span><span class="o">/</span><span class="n">rw</span><span class="o">.</span><span class="n">calc_neutral_mixing_rates</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/usage_2_GB1_9_0.png" src="../_images/usage_2_GB1_9_0.png" />
</div>
</div>
<p>We can clearly see that the relaxation times associated the first 2 components are substantially higher than the rest, suggesting that they should retain the most relevant features of the landscape. Still, faster decaying components still decay substantially slower than expectation at neutrality, possibly retaining other secondary sequence features.</p>
<p>Rendering the edges is the limiting step while plotting. Thus, for fast exploration of the landscape, it is sometimes useful skip them in the first place by not setting the <code class="docutils literal notranslate"><span class="pre">edges_df</span></code> argument, for instance, while choosing which axis to represent or the <code class="docutils literal notranslate"><span class="pre">Ns</span></code> value based on the actual visualizations and the information that they provide</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plot</span><span class="o">.</span><span class="n">init_fig</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">plot</span><span class="o">.</span><span class="n">plot_visualization</span><span class="p">(</span><span class="n">axes</span><span class="p">,</span> <span class="n">nodes_df</span><span class="p">,</span> <span class="c1">#edges_df=edges_df,</span>
                        <span class="n">edges_alpha</span><span class="o">=</span><span class="mf">0.005</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                        <span class="n">nodes_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">nodes_cmap_label</span><span class="o">=</span><span class="s1">&#39;log(Binding)&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/usage_2_GB1_11_0.png" src="../_images/usage_2_GB1_11_0.png" />
</div>
</div>
<p>The landscape seems to have 3 main fitness peaks seprated by long times to evolve from one another. Now that we have obtained a relatively reasonable visualization, we can add the edges to have a better idea about the direct connectivity among the relevant genotypes to find potential high fitness intermediates between the main peaks</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plot</span><span class="o">.</span><span class="n">init_fig</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">plot</span><span class="o">.</span><span class="n">plot_visualization</span><span class="p">(</span><span class="n">axes</span><span class="p">,</span> <span class="n">nodes_df</span><span class="p">,</span> <span class="n">edges_df</span><span class="o">=</span><span class="n">edges_df</span><span class="p">,</span>
                        <span class="n">edges_alpha</span><span class="o">=</span><span class="mf">0.005</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                        <span class="n">nodes_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">nodes_cmap_label</span><span class="o">=</span><span class="s1">&#39;log(Binding)&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/usage_2_GB1_13_0.png" src="../_images/usage_2_GB1_13_0.png" />
</div>
</div>
</section>
<section id="Characterize-the-fitness-peaks">
<h2>Characterize the fitness peaks<a class="headerlink" href="#Characterize-the-fitness-peaks" title="Permalink to this headline"></a></h2>
<p>Once we have a very general understanding of the main peaks in the landscape, we generally want to know what are the sequence determinants of those regions in sequence space.</p>
<p>One approach to do so is to highlight the alleles at each site in the visualization using the previously shown <code class="docutils literal notranslate"><span class="pre">plot.figure_allele_grid</span></code> function. However, in protein space, it is not that useful given the large number of alleles per site. An alternative approach is to roughly select genotypes belonging to each of the peaks by hand, and plotting the corresponding sequence logos easily using <code class="docutils literal notranslate"><span class="pre">logomaker</span></code></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">logomaker</span> <span class="k">as</span> <span class="nn">lm</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;39&#39;</span><span class="p">,</span> <span class="s1">&#39;40&#39;</span><span class="p">,</span> <span class="s1">&#39;41&#39;</span><span class="p">,</span> <span class="s1">&#39;54&#39;</span><span class="p">]</span>
<span class="n">peak_seqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">nodes_df</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">nodes_df</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">],</span>
             <span class="n">nodes_df</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">nodes_df</span><span class="p">[</span><span class="s1">&#39;2&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.6</span><span class="p">],</span>
             <span class="n">nodes_df</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">nodes_df</span><span class="p">[</span><span class="s1">&#39;2&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="o">-</span><span class="mf">0.75</span><span class="p">]]</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">subplots</span> <span class="o">=</span> <span class="n">plot</span><span class="o">.</span><span class="n">init_fig</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">colsize</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">rowsize</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">axes</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">peak_seqs</span><span class="p">,</span> <span class="n">subplots</span><span class="p">)):</span>
    <span class="n">m</span> <span class="o">=</span> <span class="n">lm</span><span class="o">.</span><span class="n">alignment_to_matrix</span><span class="p">(</span><span class="n">seqs</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">to_type</span><span class="o">=</span><span class="s1">&#39;probability&#39;</span><span class="p">,</span> <span class="n">pseudocount</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">m</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">logo</span> <span class="o">=</span> <span class="n">lm</span><span class="o">.</span><span class="n">Logo</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">,</span> <span class="n">color_scheme</span><span class="o">=</span><span class="s1">&#39;chemistry&#39;</span><span class="p">,</span> <span class="n">vpad</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
    <span class="n">axes</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Probability&#39;</span><span class="p">,</span> <span class="n">xlabel</span><span class="o">=</span><span class="s1">&#39;Position&#39;</span><span class="p">,</span>
                     <span class="n">xticks</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
                     <span class="n">xticklabels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Region </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/usage_2_GB1_16_0.png" src="../_images/usage_2_GB1_16_0.png" />
</div>
</div>
<p>This shows, not only that the first two positions barely contribute to the binding affinity in this assay, but also what are the sequence requirements at the last 2 positions for each peak. - The largest peak is characterized mostly by XXGX, grouping nearly 1/20 of all the possible sequences and leading to the large sequence cluster at the right hand side of the visualization. While there is a slight preference for MLF at position 54, accounting for about 50% of the sequences in the peak, we can
also observe almost any other aminoacid in this set of functional sequences. - The other two peaks, on the other hand, are characterized by a strong requirement for a specific amoniacid at position 54, and the possibility of mostly 2 variable aminoacids at 41: - Region 2 requires G at 54 and F or L at 41 - Region 3 requires A at 54 and C or A at 41</p>
</section>
<section id="Characterize-the-intermediates-between-peaks">
<h2>Characterize the intermediates between peaks<a class="headerlink" href="#Characterize-the-intermediates-between-peaks" title="Permalink to this headline"></a></h2>
<p>Besides the 3 main peaks, we can observe some clusters of sequences that connect them together for every pair of peaks. These could serve as most common intermediates to evolve from one solution to another such that mutations could only happen in a very specific order while maintaining high fitness.</p>
<p>Thus, it is also interesting to study the sequence determinants of those intermediates</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">intermediates_seqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">nodes_df</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">nodes_df</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.25</span><span class="p">,</span> <span class="n">nodes_df</span><span class="p">[</span><span class="s1">&#39;2&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="o">-</span><span class="mf">0.25</span><span class="p">)],</span>
                      <span class="n">nodes_df</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">nodes_df</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.25</span><span class="p">,</span> <span class="n">nodes_df</span><span class="p">[</span><span class="s1">&#39;2&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.25</span><span class="p">)]]</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">subplots</span> <span class="o">=</span> <span class="n">plot</span><span class="o">.</span><span class="n">init_fig</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">colsize</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">rowsize</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">axes</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">intermediates_seqs</span><span class="p">,</span> <span class="n">subplots</span><span class="p">)):</span>
    <span class="n">m</span> <span class="o">=</span> <span class="n">lm</span><span class="o">.</span><span class="n">alignment_to_matrix</span><span class="p">(</span><span class="n">seqs</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">to_type</span><span class="o">=</span><span class="s1">&#39;probability&#39;</span><span class="p">,</span> <span class="n">pseudocount</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">m</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">logo</span> <span class="o">=</span> <span class="n">lm</span><span class="o">.</span><span class="n">Logo</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">,</span> <span class="n">color_scheme</span><span class="o">=</span><span class="s1">&#39;chemistry&#39;</span><span class="p">,</span> <span class="n">vpad</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
    <span class="n">axes</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Probability&#39;</span><span class="p">,</span> <span class="n">xlabel</span><span class="o">=</span><span class="s1">&#39;Position&#39;</span><span class="p">,</span>
                     <span class="n">xticks</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
                     <span class="n">xticklabels</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;Region </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/usage_2_GB1_18_0.png" src="../_images/usage_2_GB1_18_0.png" />
</div>
</div>
<p>We can see that transitions from Region1 to regions 2 and 3 are very natural, since Region 1 is not constained at positoin 54, so it can easily take the requirements from at 54 from the other peaks without loss of fitness: XXGA and XXGG</p>
<p>On the other hand, regions 2 and 3 are more indirectly connected as at least 2 mutations are required to go from one peak to another. However, there seem to be a number of alternative ways in which these mutations could happen, leading to a variety of sequence clusters at high values in diffusion axis 1 that could work as intermediates between these two peaks</p>
</section>
<section id="Exploring-more-diffusion-axis">
<h2>Exploring more diffusion axis<a class="headerlink" href="#Exploring-more-diffusion-axis" title="Permalink to this headline"></a></h2>
<p>We can also look into more diffusion axis to try to find more fitness peaks that are not apparent from looking at 2 dimensions alone, even if the relaxation times are shorter for the following components. Lets look at diffusion axis 3, for instance</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plot</span><span class="o">.</span><span class="n">init_fig</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">plot</span><span class="o">.</span><span class="n">plot_visualization</span><span class="p">(</span><span class="n">axes</span><span class="p">,</span> <span class="n">nodes_df</span><span class="p">,</span> <span class="n">edges_df</span><span class="o">=</span><span class="n">edges_df</span><span class="p">,</span>
                        <span class="n">x</span><span class="o">=</span><span class="s1">&#39;1&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;3&#39;</span><span class="p">,</span>
                        <span class="n">edges_alpha</span><span class="o">=</span><span class="mf">0.005</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                        <span class="n">nodes_size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">nodes_cmap_label</span><span class="o">=</span><span class="s1">&#39;log(Binding)&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/usage_2_GB1_20_0.png" src="../_images/usage_2_GB1_20_0.png" />
</div>
</div>
<p>We can see how diffusion axis 3 now separates more strongly some regions in intermediate values at diffusion axis 1, generating a more or less continuous streak of high fitness sequences for which we can characterize both ends</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">streak_seqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">nodes_df</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">nodes_df</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">,</span> <span class="n">nodes_df</span><span class="p">[</span><span class="s1">&#39;3&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)],</span>
               <span class="n">nodes_df</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">logical_and</span><span class="p">(</span><span class="n">nodes_df</span><span class="p">[</span><span class="s1">&#39;1&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="o">-</span><span class="mf">0.2</span><span class="p">,</span> <span class="n">nodes_df</span><span class="p">[</span><span class="s1">&#39;3&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="o">-</span><span class="mf">0.6</span><span class="p">)]]</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">subplots</span> <span class="o">=</span> <span class="n">plot</span><span class="o">.</span><span class="n">init_fig</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">colsize</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">rowsize</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="p">(</span><span class="n">seqs</span><span class="p">,</span> <span class="n">axes</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">streak_seqs</span><span class="p">,</span> <span class="n">subplots</span><span class="p">)):</span>
    <span class="n">m</span> <span class="o">=</span> <span class="n">lm</span><span class="o">.</span><span class="n">alignment_to_matrix</span><span class="p">(</span><span class="n">seqs</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">to_type</span><span class="o">=</span><span class="s1">&#39;probability&#39;</span><span class="p">,</span> <span class="n">pseudocount</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">m</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">logo</span> <span class="o">=</span> <span class="n">lm</span><span class="o">.</span><span class="n">Logo</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">,</span> <span class="n">color_scheme</span><span class="o">=</span><span class="s1">&#39;chemistry&#39;</span><span class="p">,</span> <span class="n">vpad</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
    <span class="n">axes</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">ylabel</span><span class="o">=</span><span class="s1">&#39;Probability&#39;</span><span class="p">,</span> <span class="n">xlabel</span><span class="o">=</span><span class="s1">&#39;Position&#39;</span><span class="p">,</span>
                     <span class="n">xticks</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">m</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span>
                     <span class="n">xticklabels</span><span class="o">=</span><span class="n">labels</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/usage_2_GB1_22_0.png" src="../_images/usage_2_GB1_22_0.png" />
</div>
</div>
<ul class="simple">
<li><p>We can see that the top of the streak is characterized by a W at position 39, F/L at 41 and V, T or C at 54, while positoin 40 seems to vary quite freely.</p></li>
<li><p>The bottom of the streak is remakably different, characterized by I/V at position 49 and A at 41, while only allowing a subset of aminoacids at 54: M, S, L, C, I, F</p></li>
</ul>
</section>
<section id="References">
<h2>References<a class="headerlink" href="#References" title="Permalink to this headline"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">{bibliography}</span> <span class="pre">gpmap_tools.bib</span></code></p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="1_Ser.html" class="btn btn-neutral float-left" title="Getting started: serine landscape" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="3_ParD.html" class="btn btn-neutral float-right" title="ParD antitoxin empirical landscape" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Carlos Martí, David McCandlish.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>