<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>API Reference &mdash; Visualization of fitness landscapes 0.1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Codon models" href="usage/5_Codon.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> Visualization of fitness landscapes
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Table of Contents</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation Instructions</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage examples</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#discrete-spaces">Discrete spaces</a></li>
<li class="toctree-l2"><a class="reference internal" href="#random-walks">Random walks</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Visualization of fitness landscapes</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>API Reference</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="api-reference">
<span id="api"></span><h1>API Reference<a class="headerlink" href="#api-reference" title="Permalink to this headline"></a></h1>
<section id="discrete-spaces">
<h2>Discrete spaces<a class="headerlink" href="#discrete-spaces" title="Permalink to this headline"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="gpmap.src.space.DiscreteSpace">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">gpmap.src.space.</span></span><span class="sig-name descname"><span class="pre">DiscreteSpace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adjacency_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">state_labels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#gpmap.src.space.DiscreteSpace" title="Permalink to this definition"></a></dt>
<dd><dl class="simple">
<dt>adjacency_matrix: scipy.sparse.csr_matrix of shape (n_states, n_states)</dt><dd><p>Sparse matrix representing the adjacency relationships between
states. The ij’th entry contains a 1 if the states <cite>i</cite> and <cite>j</cite>
are connected and 0 otherwise</p>
</dd>
<dt>y: array-like of shape (n_states,)</dt><dd><p>Quantitative property associated to each state</p>
</dd>
<dt>state_labels: array-like of shape (n_genotypes, )</dt><dd><p>State labels in the sequence space</p>
</dd>
</dl>
<dl class="simple">
<dt>n_states: int</dt><dd><p>Number of states in the discrete space</p>
</dd>
<dt>state_labels: array-like of shape (n_genotypes, )</dt><dd><p>State labels in the sequence space</p>
</dd>
<dt>state_idxs: pd.Series of shape (n_genotypes, )</dt><dd><p>pd.Series containing the index of each state. It has state_labels as
index of the Series and can be used to quickly extract the index
corresponding to a set of state labels</p>
</dd>
<dt>is_regular: bool</dt><dd><p>Boolean variable storing whether the resulting graph is regular
or not, this is, whether each node has the same number of neighbors</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="gpmap.src.space.SequenceSpace">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">gpmap.src.space.</span></span><span class="sig-name descname"><span class="pre">SequenceSpace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_alleles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'dna'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#gpmap.src.space.SequenceSpace" title="Permalink to this definition"></a></dt>
<dd><p>Class for creating a Sequence space characterized by having sequences as
states. States are connected in the discrete space if they differ by a 
single position in the sequence. It can be created in two different ways:</p>
<blockquote>
<div><ul class="simple">
<li><p>From a set of sequences and function values X, y</p></li>
<li><p>By specifying the properties of the sequence space (alphabet,
sequence length, number of alleles per site and type of alphabet)</p></li>
</ul>
</div></blockquote>
<dl class="simple">
<dt>X: array-like of shape (n_genotypes,)</dt><dd><p>Sequences to use as state labels of the discrete sequence space</p>
</dd>
<dt>y: array-like of shape (n_genotypes,)</dt><dd><p>Quantitative phenotype or fitness associated to each genotype</p>
</dd>
<dt>seq_length: int (None)</dt><dd><p>Length of the sequences in the sequence space. If not given, it will be
guessed from <cite>alphabet</cite> or <cite>n_alleles</cite></p>
</dd>
<dt>n_alleles: list of size <cite>seq_length</cite> (None)</dt><dd><p>List containing the number of alleles present in each of the sites
of the sequence space. It can only be specified for 
<cite>alphabet_type=custom</cite></p>
</dd>
<dt>alphabet_type: str (‘dna’)</dt><dd><p>Sequence type: {‘dna’, ‘rna’, ‘protein’, ‘custom’}</p>
</dd>
<dt>alphabet: list of <a href="#id1"><span class="problematic" id="id2">`</span></a>seq_length’ lists</dt><dd><p>Every element of the list is itself a list containing the different
alleles allowed in each site. Note that the number and type of alleles
can be different for every site.</p>
</dd>
</dl>
<dl class="simple">
<dt>n_genotypes: int</dt><dd><p>Number of states in the complete sequence space</p>
</dd>
<dt>genotypes: array-like of shape (n_genotypes, )</dt><dd><p>Genotype labels in the sequence space</p>
</dd>
<dt>adjacency_matrix: scipy.sparse.csr_matrix of shape (n_genotypes, n_genotypes)</dt><dd><p>Sparse matrix representing the adjacency relationships between
genotypes. The ij’th entry contains a 1 if the genotypes <cite>i</cite> and <cite>j</cite>
are separated by a single mutation and 0 otherwise</p>
</dd>
<dt>y: array-like of shape (n_genotypes,)</dt><dd><p>Quantitative phenotype or fitness associated to each genotype</p>
</dd>
<dt>is_regular: bool</dt><dd><p>Boolean variable storing whether the resulting Hamming graph is regular
or not. In other words, whether every site has the same number of
alleles</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="gpmap.src.space.SequenceSpace.remove_codon_incompatible_transitions">
<span class="sig-name descname"><span class="pre">remove_codon_incompatible_transitions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">codon_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Standard'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#gpmap.src.space.SequenceSpace.remove_codon_incompatible_transitions" title="Permalink to this definition"></a></dt>
<dd><p>Recalculates the adjacency matrix of the discrete space to only allow
transitions that are compatible with the specified codon table</p>
<dl class="simple">
<dt>codon_table: str or Bio.Data.CodonTable</dt><dd><p>NCBI code for an existing genetic code or a custom CodonTable 
object to translate nucleotide sequences into protein</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="gpmap.src.space.SequenceSpace.to_nucleotide_space">
<span class="sig-name descname"><span class="pre">to_nucleotide_space</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">codon_table</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Standard'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_y</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alphabet_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'dna'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#gpmap.src.space.SequenceSpace.to_nucleotide_space" title="Permalink to this definition"></a></dt>
<dd><p>Transforms a protein space into a nucleotide space using a codon table
for translating the sequence</p>
<dl class="simple">
<dt>codon_table: str or Bio.Data.CodonTable</dt><dd><p>NCBI code for an existing genetic code or a custom CodonTable 
object to translate nucleotide sequences into protein</p>
</dd>
<dt>stop_y: float (None)</dt><dd><p>Value of the function given for every nucleotide sequence with an
in-frame stop codon. If ‘None’, it will use the minimum
value found across of the sequences, assumed to be equal to a 
complete loss of function</p>
</dd>
<dt>alphabet_type: str (‘dna’)</dt><dd><p>Sequence type to use in the resulting nucleotide space
It can only take one of the following values {‘dna’, ‘rna’}</p>
</dd>
</dl>
<dl class="simple">
<dt>SequenceSpace</dt><dd><p>Nucleotide sequence space with 4 alleles per site and 3 times 
the number of sites of the current space</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="random-walks">
<h2>Random walks<a class="headerlink" href="#random-walks" title="Permalink to this headline"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="gpmap.src.randwalk.TimeReversibleRandomWalk">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">gpmap.src.randwalk.</span></span><span class="sig-name descname"><span class="pre">TimeReversibleRandomWalk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">space</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#gpmap.src.randwalk.TimeReversibleRandomWalk" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="gpmap.src.randwalk.WMWSWalk">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">gpmap.src.randwalk.</span></span><span class="sig-name descname"><span class="pre">WMWSWalk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">space</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#gpmap.src.randwalk.WMWSWalk" title="Permalink to this definition"></a></dt>
<dd><p>Class for Weak Mutation Weak Selection Random Walk on a SequenceSpace</p>
<p>…</p>
<dl class="simple">
<dt>space<span class="classifier">DiscreteSpace class</span></dt><dd><p>Space on which the random walk takes place</p>
</dd>
<dt>Ns<span class="classifier">real </span></dt><dd><p>Scaled effective population size for the evolutionary model</p>
</dd>
<dt>rate_matrix<span class="classifier">csr_matrix</span></dt><dd><p>Rate matrix defining the continuous time process</p>
</dd>
</dl>
<dl class="simple">
<dt>set_Ns():</dt><dd><p>Method to specify the scaled effective population size Ns, either directly
or by specifying the mean function at stationarity or the percentile
it represents from the distribution of functions across sequence space</p>
</dd>
<dt>calc_stationary_frequencies():</dt><dd><p>Calculates the stationary frequencies of the genotypes under the random
walk specified on the discrete space</p>
</dd>
<dt>calc_rate_matrix():</dt><dd><p>Calculates the rate matrix for the continuous time process given 
the scaled effective population size (Ns) or average phenotype at
stationarity.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="gpmap.src.randwalk.WMWSWalk.calc_neutral_mixing_rates">
<span class="sig-name descname"><span class="pre">calc_neutral_mixing_rates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">neutral_site_Qs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">neutral_stat_freqs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">site_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#gpmap.src.randwalk.WMWSWalk.calc_neutral_mixing_rates" title="Permalink to this definition"></a></dt>
<dd><p>Calculates the neutral mixing rates for a SequenceSpace
In case no GTR mutation model is specified, then the neutral
mixing rates is limited by the site with the least number of alleles.
Otherwise, as we assume that mutations are site-independent, 
the slowest neutral mixing rate is going to by limited by the slowest
site, provided by the smallest of second eigenvalues in the site
rate matrices</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="gpmap.src.randwalk.WMWSWalk.calc_stationary_frequencies">
<span class="sig-name descname"><span class="pre">calc_stationary_frequencies</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#gpmap.src.randwalk.WMWSWalk.calc_stationary_frequencies" title="Permalink to this definition"></a></dt>
<dd><p>Calculates the genotype stationary frequencies using Ns stored in 
the object and stores the corresponding diagonal matrices with the
sqrt transformation and its inverse</p>
</dd></dl>

</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="usage/5_Codon.html" class="btn btn-neutral float-left" title="Codon models" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Carlos Martí, David McCandlish.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>