


<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xml:lang="" lang="" version="-//W3C//DTD XHTML 1.1//EN" xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">

  
  <title>&lt;no title&gt; &mdash; Visualization of fitness landscapes 0.1.0 documentation</title>
  

  

  

    <link href='https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'/>
    <link rel="stylesheet" href="_static/css/pdj.css" type="text/css" />

  
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  
    <link rel="stylesheet" href="_static/css/pdj.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="Visualization of fitness landscapes 0.1.0 documentation" href="index.html"/>
        <link rel="prev" title="Visualization of fitness landscapes" href="index.html"/>

    <meta http-equiv="content-type" content="text/html; charset=utf-8" />
    <meta http-equiv="cache-control" content="public" />
    <meta name="robots" content="follow, all" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <!-- Add jQuery library -->
    <script type="text/javascript" src="http://code.jquery.com/jquery-latest.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

  </head>

  <body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        <a href="index.html" class="fa fa-home"> Visualization of fitness landscapes </a>
        <div role="search">
	  <form id ="rtd-search-form" class="wy-form"
		action="search.html" method="get">
	    <input type="text" name="q" placeholder="Search docs" />
	    <input type="hidden" name="check_keywords" value="yes" />
	    <input type="hidden" name="area" value="default" />
	  </form>
	</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	
          
          
              <!-- Local TOC -->
              <div class="local-toc"><ul class="simple">
</ul>
</div>
          
        

      </div>
      &nbsp;
    </nav>
    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      <nav class="wy-nav-top" id="barra-mobile" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="#">Porão do Juca</a>
      </nav>

      <div class="wy-nav-content">
	<div class="fundo-claro">
	</div>
	<div class="fundo-escuro">
	</div>

        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
	    
	    <!-- <ul class="wy-breadcrumbs"> -->
	    <!--   <li><a href="#">Docs</a> &raquo;</li> -->

	    <!--   <li>Features</li> -->
	    <!--   <li class="wy-breadcrumbs-aside"> -->

	    <!-- 	<a href="_sources/index.txt" rel="nofollow"> View page source</a> -->

	    <!--   </li> -->
	    <!-- </ul> -->
	    <!-- <hr/> -->
	  </div>

          <div role="main" class="">

	    <div id="content" class="hfeed entry-container hentry">
  <dl>
<dt>{</dt><dd><dl>
<dt>“cells”: [</dt><dd><dl>
<dt>{</dt><dd><p>“cell_type”: “markdown”,
“id”: “a567e30b”,
“metadata”: {},
“source”: [</p>
<blockquote>
<div><p>“# Serine codon landscapen”,
“n”,
“Serine is the only aminoacid that is encoded by two sets of mutationally disconnected codons under the standard genetic code.n”,
“n”,
“- AGY: {AGU,AGC}n”,
“- UCN: {UCA,UCC,UCG,UCU}n”,
“n”,
“This leads to a landscape with two isolated fitness peaks that we can easily visualizen”,
“n”,
“## 1. Defining the discrete spacen”,
“n”,
“The first thing we need to do is to define the discrete space for the evolutionary random walk. While we provide a generic class DiscreteSpace to define an arbitrary discrete space based on the adjacency matrix and nodes properties on which the transition between states may depend, we are going to use the class SequenceSpace that has specific built-in properties and methods specifically for sequence space.”</p>
</div></blockquote>
<p>]</p>
</dd>
</dl>
<p>},
{</p>
<blockquote>
<div><p>“cell_type”: “code”,
“execution_count”: 1,
“id”: “11574b74”,
“metadata”: {},
“outputs”: [],
“source”: [</p>
<blockquote>
<div><p>“# Import required librariesn”,
“import pandas as pdn”,
“from gpmap.src.space import SequenceSpace”</p>
</div></blockquote>
<p>]</p>
</div></blockquote>
<p>},
{</p>
<blockquote>
<div><p>“cell_type”: “markdown”,
“id”: “2ea705dd”,
“metadata”: {},
“source”: [</p>
<blockquote>
<div><p>“After importing the required libraries, we can read the fitness values that we have previously generated. In this artificial example, we assigned fitnesses in the following way:n”,
“n”,
“- w=2 to codons encoding Serinen”,
“- w=1 to codons encoding other aminoacidesn”,
“- w=0 to stop codonsn”,
“n”,
“We also added some small perturbation to the fitnesses of the individual codons that could account for codon usage biases, but also allow better separation of the genotypes in the low dimensional representation. We provide different ways to generate this particular landscapen”,
“n”,
“### Directly from function datan”</p>
</div></blockquote>
<p>]</p>
</div></blockquote>
<p>},
{</p>
<blockquote>
<div><p>“cell_type”: “code”,
“execution_count”: 3,
“id”: “81778ec8”,
“metadata”: {},
“outputs”: [</p>
<blockquote>
<div><dl>
<dt>{</dt><dd><dl>
<dt>“data”: {</dt><dd><dl class="simple">
<dt>“text/html”: [</dt><dd><p>“&lt;div&gt;n”,
“&lt;style scoped&gt;n”,
”    .dataframe tbody tr th:only-of-type {n”,
”        vertical-align: middle;n”,
”    }n”,
“n”,
”    .dataframe tbody tr th {n”,
”        vertical-align: top;n”,
”    }n”,
“n”,
”    .dataframe thead th {n”,
”        text-align: right;n”,
”    }n”,
“&lt;/style&gt;n”,
“&lt;table border=&quot;1&quot; class=&quot;dataframe&quot;&gt;n”,
”  &lt;thead&gt;n”,
”    &lt;tr style=&quot;text-align: right;&quot;&gt;n”,
”      &lt;th&gt;&lt;/th&gt;n”,
”      &lt;th&gt;function&lt;/th&gt;n”,
”    &lt;/tr&gt;n”,
”  &lt;/thead&gt;n”,
”  &lt;tbody&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;AAA&lt;/th&gt;n”,
”      &lt;td&gt;1.176405&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;AAC&lt;/th&gt;n”,
”      &lt;td&gt;1.040016&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;AAG&lt;/th&gt;n”,
”      &lt;td&gt;1.097874&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;AAU&lt;/th&gt;n”,
”      &lt;td&gt;1.224089&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;ACA&lt;/th&gt;n”,
”      &lt;td&gt;1.186756&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”  &lt;/tbody&gt;n”,
“&lt;/table&gt;n”,
“&lt;/div&gt;”</p>
</dd>
</dl>
<p>],
“text/plain”: [</p>
<blockquote>
<div><p>”     functionn”,
“AAA  1.176405n”,
“AAC  1.040016n”,
“AAG  1.097874n”,
“AAU  1.224089n”,
“ACA  1.186756”</p>
</div></blockquote>
<p>]</p>
</dd>
</dl>
<p>},
“execution_count”: 3,
“metadata”: {},
“output_type”: “execute_result”</p>
</dd>
</dl>
<p>}</p>
</div></blockquote>
<p>],
“source”: [</p>
<blockquote>
<div><p>“fpath = ‘../gpmap/test/data/serine.csv’n”,
“data = pd.read_csv(fpath, index_col=0)n”,
“data.head()”</p>
</div></blockquote>
<p>]</p>
</div></blockquote>
<p>},
{</p>
<blockquote>
<div><p>“cell_type”: “markdown”,
“id”: “885e81e8”,
“metadata”: {},
“source”: [</p>
<blockquote>
<div><p>“We can see the simple table that just stores the function value for each sequence, that we will use to create a SequenceSpace object, that has a number of attributes such as number of states, allelese or the alphabet_type, in this case ‘rna’”</p>
</div></blockquote>
<p>]</p>
</div></blockquote>
<p>},
{</p>
<blockquote>
<div><p>“cell_type”: “code”,
“execution_count”: 26,
“id”: “012e5a44”,
“metadata”: {},
“outputs”: [</p>
<blockquote>
<div><dl>
<dt>{</dt><dd><dl>
<dt>“data”: {</dt><dd><dl class="simple">
<dt>“text/plain”: [</dt><dd><p>“(64, [4, 4, 4], ‘rna’)”</p>
</dd>
</dl>
<p>]</p>
</dd>
</dl>
<p>},
“execution_count”: 26,
“metadata”: {},
“output_type”: “execute_result”</p>
</dd>
</dl>
<p>}</p>
</div></blockquote>
<p>],
“source”: [</p>
<blockquote>
<div><p>“space = SequenceSpace(seq_length=3, alphabet_type=’rna’, function=data[‘function’])n”,
“space.n_states, space.n_alleles, space.alphabet_type”</p>
</div></blockquote>
<p>]</p>
</div></blockquote>
<p>},
{</p>
<blockquote>
<div><p>“cell_type”: “markdown”,
“id”: “73ef298b”,
“metadata”: {},
“source”: [</p>
<blockquote>
<div><p>“### Using codon modeln”,
“n”,
“Sometimes we may not be able to differentiate between the function or fitnesses of different codons encoding the same aminoacid, but still want to take into account the connectivity at the nucleotide level for visualizing the landscape as in a codon model of evolution.n”,
“n”,
“The following table contains the fitnesses associated to each of the 20 aminoacids”</p>
</div></blockquote>
<p>]</p>
</div></blockquote>
<p>},
{</p>
<blockquote>
<div><p>“cell_type”: “code”,
“execution_count”: 27,
“id”: “251dcb04”,
“metadata”: {},
“outputs”: [</p>
<blockquote>
<div><dl>
<dt>{</dt><dd><dl>
<dt>“data”: {</dt><dd><dl class="simple">
<dt>“text/html”: [</dt><dd><p>“&lt;div&gt;n”,
“&lt;style scoped&gt;n”,
”    .dataframe tbody tr th:only-of-type {n”,
”        vertical-align: middle;n”,
”    }n”,
“n”,
”    .dataframe tbody tr th {n”,
”        vertical-align: top;n”,
”    }n”,
“n”,
”    .dataframe thead th {n”,
”        text-align: right;n”,
”    }n”,
“&lt;/style&gt;n”,
“&lt;table border=&quot;1&quot; class=&quot;dataframe&quot;&gt;n”,
”  &lt;thead&gt;n”,
”    &lt;tr style=&quot;text-align: right;&quot;&gt;n”,
”      &lt;th&gt;&lt;/th&gt;n”,
”      &lt;th&gt;function&lt;/th&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;protein&lt;/th&gt;n”,
”      &lt;th&gt;&lt;/th&gt;n”,
”    &lt;/tr&gt;n”,
”  &lt;/thead&gt;n”,
”  &lt;tbody&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;A&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;C&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;D&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;E&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;F&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;G&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;H&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;I&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;K&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;L&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;M&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;N&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;P&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;Q&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;R&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;S&lt;/th&gt;n”,
”      &lt;td&gt;2&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;T&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;V&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;W&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”    &lt;tr&gt;n”,
”      &lt;th&gt;Y&lt;/th&gt;n”,
”      &lt;td&gt;1&lt;/td&gt;n”,
”    &lt;/tr&gt;n”,
”  &lt;/tbody&gt;n”,
“&lt;/table&gt;n”,
“&lt;/div&gt;”</p>
</dd>
</dl>
<p>],
“text/plain”: [</p>
<blockquote>
<div><p>”         functionn”,
“protein          n”,
“A               1n”,
“C               1n”,
“D               1n”,
“E               1n”,
“F               1n”,
“G               1n”,
“H               1n”,
“I               1n”,
“K               1n”,
“L               1n”,
“M               1n”,
“N               1n”,
“P               1n”,
“Q               1n”,
“R               1n”,
“S               2n”,
“T               1n”,
“V               1n”,
“W               1n”,
“Y               1”</p>
</div></blockquote>
<p>]</p>
</dd>
</dl>
<p>},
“execution_count”: 27,
“metadata”: {},
“output_type”: “execute_result”</p>
</dd>
</dl>
<p>}</p>
</div></blockquote>
<p>],
“source”: [</p>
<blockquote>
<div><p>“protein_data = pd.read_csv(‘../gpmap/test/data/serine.protein.csv’, index_col=0)n”,
“protein_data”</p>
</div></blockquote>
<p>]</p>
</div></blockquote>
<p>},
{</p>
<blockquote>
<div><p>“cell_type”: “code”,
“execution_count”: 28,
“id”: “a5874a8c”,
“metadata”: {},
“outputs”: [</p>
<blockquote>
<div><dl>
<dt>{</dt><dd><dl>
<dt>“data”: {</dt><dd><dl class="simple">
<dt>“text/plain”: [</dt><dd><p>“(64, [4, 4, 4], ‘rna’)”</p>
</dd>
</dl>
<p>]</p>
</dd>
</dl>
<p>},
“execution_count”: 28,
“metadata”: {},
“output_type”: “execute_result”</p>
</dd>
</dl>
<p>}</p>
</div></blockquote>
<p>],
“source”: [</p>
<blockquote>
<div><p>“space = SequenceSpace(seq_length=3, alphabet_type=’rna’, function=protein_data[‘function’],n”,
”                      codon_table=’Standard’, stop_function=0)n”,
“space.n_states, space.n_alleles, space.alphabet_type”</p>
</div></blockquote>
<p>]</p>
</div></blockquote>
<p>},
{</p>
<blockquote>
<div><p>“cell_type”: “markdown”,
“id”: “b5661ba3”,
“metadata”: {},
“source”: [</p>
<blockquote>
<div><p>“### Using CodonSpace classn”,
“n”,
“We also provide a more generic CodonSpace class that does this operation for us so that we only need to provide the aminoacid(s) are are going to be under selection, enabling also to visualizing the structure of the landscape corresponding to aminoacids with certain properties”</p>
</div></blockquote>
<p>]</p>
</div></blockquote>
<p>},
{</p>
<blockquote>
<div><p>“cell_type”: “code”,
“execution_count”: 29,
“id”: “5a49d5d4”,
“metadata”: {},
“outputs”: [],
“source”: [</p>
<blockquote>
<div><p>“from gpmap.src.space import CodonSpace”</p>
</div></blockquote>
<p>]</p>
</div></blockquote>
<p>},
{</p>
<blockquote>
<div><p>“cell_type”: “code”,
“execution_count”: 31,
“id”: “3ad810a1”,
“metadata”: {},
“outputs”: [</p>
<blockquote>
<div><dl>
<dt>{</dt><dd><dl>
<dt>“data”: {</dt><dd><dl class="simple">
<dt>“text/plain”: [</dt><dd><p>“(64, [4, 4, 4], ‘rna’)”</p>
</dd>
</dl>
<p>]</p>
</dd>
</dl>
<p>},
“execution_count”: 31,
“metadata”: {},
“output_type”: “execute_result”</p>
</dd>
</dl>
<p>}</p>
</div></blockquote>
<p>],
“source”: [</p>
<blockquote>
<div><p>“space = CodonSpace(allowed_aminoacids=[‘S’], codon_table=’Standard’, add_variation=True, seed=0)n”,
“space.n_states, space.n_alleles, space.alphabet_type”</p>
</div></blockquote>
<p>]</p>
</div></blockquote>
<p>},
{</p>
<blockquote>
<div><p>“cell_type”: “markdown”,
“id”: “893e5a64”,
“metadata”: {},
“source”: [</p>
<blockquote>
<div><p>“Note that we could also test how these landscapes would change under different genetic codes other than the standard. We use biopython module to translate the nucleotide sequence into protein sequence using [NCBI reference](<a class="reference external" href="https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi">https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi</a>) for different codon tables or genetic codesn”,
“n”,
“## 2. Defining the random walk in the discrete sequence spacen”,
“n”</p>
</div></blockquote>
<p>]</p>
</div></blockquote>
<p>}</p>
</dd>
</dl>
<p>],
“metadata”: {</p>
<blockquote>
<div><dl class="simple">
<dt>“kernelspec”: {</dt><dd><p>“display_name”: “Python 3”,
“language”: “python”,
“name”: “python3”</p>
</dd>
</dl>
<p>},
“language_info”: {</p>
<blockquote>
<div><dl class="simple">
<dt>“codemirror_mode”: {</dt><dd><p>“name”: “ipython”,
“version”: 3</p>
</dd>
</dl>
<p>},
“file_extension”: “.py”,
“mimetype”: “text/x-python”,
“name”: “python”,
“nbconvert_exporter”: “python”,
“pygments_lexer”: “ipython3”,
“version”: “3.6.13”</p>
</div></blockquote>
<p>}</p>
</div></blockquote>
<p>},
“nbformat”: 4,
“nbformat_minor”: 5</p>
</dd>
</dl>
<p>}</p>


	    </div>
            <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="index.html" class="btn btn-neutral" title="Visualization of fitness landscapes"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2022, Carlos Martí, David McCandlish.
    </p>
  </div>

  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/jucacrispim/sphinx_pdj_theme">theme</a> provided by <a href="http://poraodojuca.net">Porão do Juca</a>.

</footer>
	</div>
	</div>
	  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.1.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
    <script type="text/javascript" src="_static/documentation_options.js">

    </script>
    <script type="text/javascript" src="_static/jquery.js">

    </script>
    <script type="text/javascript" src="_static/underscore.js">

    </script>
    <script type="text/javascript" src="_static/doctools.js">

    </script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js">

    </script>

  

   <script type="text/javascript"
           src="_static/js/theme.js"></script>

   <script type="text/javascript"
           src="_static/js/pdj.js"></script>

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

  </body>
</html>